### PR TITLE
fix #13837 the Invalid log : worker worker have not received the hear…

### LINKED ISSUE #13837: [Bug] [Master] worker worker have not received the heartbeat
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

The `worker` worker Is inconsistent with the ip addr requirement.

```
2023-03-31 00:40:51 | [WARN] 2023-03-31 00:40:51.139 org.apache.dolphinscheduler.server.master.dispatch.host.LowerWeightHostManager:[132] - [WorkflowInstance-0][TaskInstance-0] - worker worker in work group default have not received the heartbeat
```
and a another key log like this  
```
[INFO] 2023-03-27 19:47:47.186 org.apache.dolphinscheduler.server.master.registry.ServerNodeManager:[197] - [WorkflowInstance-0][TaskInstance-0] - Worker: /nodes/worker added, currentNode : worker
```

### What you expected to happen

There are no invalid WARN logs

### How to reproduce

A started standalone cluster.

### Anything else

_No response_

### Version

dev

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by lxorc**
fix #13837 the Invalid log : worker worker have not received the hearbet

### CODE REVIEW

**Comment by ruanwenjun** [comment]
We don't need to add this check, this case might be due to we registering an empty node? Could you please describe how to reproduce this.

**Comment by lxorc** [comment]
> We don't need to add this check, this case might be due to we registering an empty node? Could you please describe how to reproduce this.

Yes, by default a `worker` node will be loaded in `registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); `

```java
registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); 
String[] parts = path.split("/");  // REGISTRY_DOLPHINSCHEDULER_WORKERS = /nodes/worker
final String workerAddress = parts[parts.length - 1];  // workerAddress  = worker
if (type == Type.UPDATE) // the Type.UPDATE is  Node /nodes/worker update ,   case NODE_UPDATED: type(Type.UPDATE);
syncSingleWorkerNodeInfo(workerAddress, JSONUtils.parseObject(data, WorkerHeartBeat.class));  //  ???
workerNodeInfo.put(workerAddress, info);  // put "worker" key into the workerNodeInfo
```


The effect is that the `addr` of `getHostWeight` will be the `worker` and `heartBeat == null`.

so 

```java
        if (heartBeat == null) {
            logger.warn("worker {} in work group {} have not received the heartbeat", addr, workerGroup);
            return Optional.empty();
        }
```

**Comment by ruanwenjun** [comment]
> > We don't need to add this check, this case might be due to we registering an empty node? Could you please describe how to reproduce this.
> 
> Yes, by default a `worker` node will be loaded in `registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); `
> 
> ```java
> registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); 
> String[] parts = path.split("/");  // REGISTRY_DOLPHINSCHEDULER_WORKERS = /nodes/worker
> final String workerAddress = parts[parts.length - 1];  // workerAddress  = worker
> if (type == Type.UPDATE) // the Type.UPDATE is  Node /nodes/worker update ,   case NODE_UPDATED: type(Type.UPDATE);
> syncSingleWorkerNodeInfo(workerAddress, JSONUtils.parseObject(data, WorkerHeartBeat.class));  //  ???
> workerNodeInfo.put(workerAddress, info);  // put "worker" key into the workerNodeInfo
> ```
> 
> The effect is that the `addr` of `getHostWeight` will be the `worker` and `heartBeat == null`.
> 
> so
> 
> ```java
>         if (heartBeat == null) {
>             logger.warn("worker {} in work group {} have not received the heartbeat", addr, workerGroup);
>             return Optional.empty();
>         }
> ```

So we only need to subscribe the event of REGISTRY_DOLPHINSCHEDULER_WORKERS's child path

**Comment by lxorc** [comment]
> > > We don't need to add this check, this case might be due to we registering an empty node? Could you please describe how to reproduce this.
> > 
> > 
> > Yes, by default a `worker` node will be loaded in `registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); `
> > ```java
> > registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); 
> > String[] parts = path.split("/");  // REGISTRY_DOLPHINSCHEDULER_WORKERS = /nodes/worker
> > final String workerAddress = parts[parts.length - 1];  // workerAddress  = worker
> > if (type == Type.UPDATE) // the Type.UPDATE is  Node /nodes/worker update ,   case NODE_UPDATED: type(Type.UPDATE);
> > syncSingleWorkerNodeInfo(workerAddress, JSONUtils.parseObject(data, WorkerHeartBeat.class));  //  ???
> > workerNodeInfo.put(workerAddress, info);  // put "worker" key into the workerNodeInfo
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The effect is that the `addr` of `getHostWeight` will be the `worker` and `heartBeat == null`.
> > so
> > ```java
> >         if (heartBeat == null) {
> >             logger.warn("worker {} in work group {} have not received the heartbeat", addr, workerGroup);
> >             return Optional.empty();
> >         }
> > ```
> 
> So we only need to subscribe the event of REGISTRY_DOLPHINSCHEDULER_WORKERS's child path

Please help to review my code to see if there are other problems. If there is a better implementation, please reply to me. And I have fixed it in my production environment and it works fine.

**Comment by ruanwenjun** [comment]
> > > > We don't need to add this check, this case might be due to we registering an empty node? Could you please describe how to reproduce this.
> > > 
> > > 
> > > Yes, by default a `worker` node will be loaded in `registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); `
> > > ```java
> > > registryClient.subscribe(REGISTRY_DOLPHINSCHEDULER_WORKERS, new WorkerDataListener()); 
> > > String[] parts = path.split("/");  // REGISTRY_DOLPHINSCHEDULER_WORKERS = /nodes/worker
> > > final String workerAddress = parts[parts.length - 1];  // workerAddress  = worker
> > > if (type == Type.UPDATE) // the Type.UPDATE is  Node /nodes/worker update ,   case NODE_UPDATED: type(Type.UPDATE);
> > > syncSingleWorkerNodeInfo(workerAddress, JSONUtils.parseObject(data, WorkerHeartBeat.class));  //  ???
> > > workerNodeInfo.put(workerAddress, info);  // put "worker" key into the workerNodeInfo
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The effect is that the `addr` of `getHostWeight` will be the `worker` and `heartBeat == null`.
> > > so
> > > ```java
> > >         if (heartBeat == null) {
> > >             logger.warn("worker {} in work group {} have not received the heartbeat", addr, workerGroup);
> > >             return Optional.empty();
> > >         }
> > > ```
> > 
> > 
> > So we only need to subscribe the event of REGISTRY_DOLPHINSCHEDULER_WORKERS's child path
> 
> Please help to review my code to see if there are other problems. If there is a better implementation, please reply to me. And I have fixed it in my production environment and it works fine.

Does master path will have this problem? could you please change the master together?

**Comment by ruanwenjun** [comment]
Please fix the code style by `mvn spotless:apply`

File: dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/RegistryClient.java [hunk 1]
```diff
@@ -210,7 +210,7 @@ public boolean isMasterPath(String path) {
     }
 
     public boolean isWorkerPath(String path) {
-        return path != null && path.startsWith(RegistryNodeType.WORKER.getRegistryPath());
+        return path != null && path.startsWith(Constants.REGISTRY_DOLPHINSCHEDULER_WORKERS) && !path.equals(Constants.REGISTRY_DOLPHINSCHEDULER_WORKERS);
```
Comment (Gallardot) [inline dolphinscheduler-registry/dolphinscheduler-registry-api/src/main/java/org/apache/dolphinscheduler/registry/api/RegistryClient.java]:
```suggestion
        return path != null && path.startsWith(Constants.REGISTRY_DOLPHINSCHEDULER_WORKERS)
                && !path.equals(Constants.REGISTRY_DOLPHINSCHEDULER_WORKERS);
```


@lxorc Do you have time to work on this PR? Maybe I can help with that. We also encountered this issue, and I have tested this PR in our environment. It is feasible.

**Review by Gallardot — CHANGES_REQUESTED** [review]

**Pull request closed without merging**